File: dbgen/p06_saas.py
import duckdb, numpy as np, sys, os
from datetime import datetime, timedelta, date
from concurrent.futures import ProcessPoolExecutor
from utils.synth_utils import (
    GenerationProgress,
    batched_insert,
    get_worker_count,
    print_generation_summary,
    run_parallel,
)


def generate_accounts_chunk(start, end, industries, base):
    size = end - start
    rng = np.random.default_rng(start)
    industry_indices = rng.integers(0, len(industries), size)
    countries = ["US", "UK", "DE", "FR", "CA", "AU"]
    country_indices = rng.integers(0, len(countries), size)
    arrs = rng.uniform(5000, 500000, size)
    days_offset = rng.integers(0, 701, size)
    csm_ids = rng.integers(1, 21, size)
    health_scores = rng.integers(1, 101, size)
    
    account_ids = range(start, end)
    account_names = [f"Account {i}" for i in account_ids]
    selected_industries = np.take(industries, industry_indices).tolist()
    selected_countries = np.take(countries, country_indices).tolist()
    arrs_rounded = np.round(arrs, 2).tolist()
    created_dates = (np.datetime64(base) + days_offset.astype("timedelta64[D]")).tolist()
    selected_csm_ids = csm_ids.tolist()
    selected_health_scores = health_scores.tolist()
    
    return list(zip(account_ids, account_names, selected_industries, selected_countries, arrs_rounded, created_dates, selected_csm_ids, selected_health_scores))


def generate_subscriptions_chunk(start, end, NAC, plans, base):
    size = end - start
    rng = np.random.default_rng(start)
    account_ids = rng.integers(1, NAC + 1, size)
    plan_indices = rng.integers(0, len(plans), size)
    seats = rng.integers(1, 201, size)
    mrrs = rng.uniform(99, 9999, size)
    days_offset = rng.integers(0, 601, size)
    active_probs = rng.random(size)

    sub_ids = range(start, end)
    selected_account_ids = account_ids.tolist()
    selected_plans = np.take(plans, plan_indices).tolist()
    selected_seats = seats.tolist()
    mrrs_rounded = np.round(mrrs, 2).tolist()
    
    start_dates = (np.datetime64(base) + days_offset.astype("timedelta64[D]"))
    end_dates = start_dates + np.timedelta64(365, 'D')
    
    is_active = (active_probs > 0.1).tolist()
    
    return list(zip(
        sub_ids,
        selected_account_ids,
        selected_plans,
        selected_seats,
        mrrs_rounded,
        start_dates.tolist(),
        end_dates.tolist(),
        is_active,
        end_dates.tolist()
    ))


def generate_events_chunk(start, end, NAC, etypes, bts):
    size = end - start
    rng = np.random.default_rng(start)
    account_ids = rng.integers(1, NAC + 1, size)
    user_ids = rng.integers(1, NAC * 5 + 1, size)
    etype_indices = rng.integers(0, len(etypes), size)
    seconds_offset = rng.integers(0, 700 * 86400 + 1, size)
    session_ids_raw = rng.integers(1, NAC * 20 + 1, size)
    platforms = ["web", "mobile", "api"]
    platform_indices = rng.integers(0, len(platforms), size)

    event_ids = range(start, end)
    selected_account_ids = account_ids.tolist()
    selected_user_ids = user_ids.tolist()
    selected_etypes = np.take(etypes, etype_indices).tolist()
    event_tss = (np.datetime64(bts) + seconds_offset.astype("timedelta64[s]")).tolist()
    session_ids = [f"sess_{sid}" for sid in session_ids_raw]
    selected_platforms = np.take(platforms, platform_indices).tolist()
    
    return list(zip(event_ids, selected_account_ids, selected_user_ids, selected_etypes, event_tss, session_ids, selected_platforms))


def generate_feature_usage_chunk(start, end, NAC, features, base):
    size = end - start
    rng = np.random.default_rng(start)
    account_ids = rng.integers(1, NAC + 1, size)
    feature_indices = rng.integers(0, len(features), size)
    days_offset = rng.integers(0, 701, size)
    usage_counts = rng.integers(1, 1001, size)

    fu_ids = range(start, end)
    selected_account_ids = account_ids.tolist()
    selected_features = np.take(features, feature_indices).tolist()
    usage_dates = (np.datetime64(base) + days_offset.astype("timedelta64[D]")).tolist()
    selected_usage_counts = usage_counts.tolist()
    
    return list(zip(fu_ids, selected_account_ids, selected_features, usage_dates, selected_usage_counts))


def generate_support_tickets_chunk(start, end, NAC, bts, priorities, ticket_cats):
    size = end - start
    rng = np.random.default_rng(start)
    account_ids = rng.integers(1, NAC + 1, size)
    created_offsets = rng.integers(0, 700 * 86400 + 1, size)
    resolved_offsets = rng.integers(0, 700 * 86400 + 1, size)
    resolved_probs = rng.random(size)
    priority_indices = rng.integers(0, len(priorities), size)
    cat_indices = rng.integers(0, len(ticket_cats), size)
    csat_scores = rng.integers(1, 6, size)
    csat_probs = rng.random(size)
    is_resolved_probs = rng.random(size)

    ticket_ids = range(start, end)
    selected_account_ids = account_ids.tolist()
    created_tss = (np.datetime64(bts) + created_offsets.astype("timedelta64[s]")).tolist()
    
    resolved_tss = [
        (np.datetime64(bts) + np.timedelta64(off, 's')).tolist() if prob > 0.2 else None 
        for off, prob in zip(resolved_offsets, resolved_probs)
    ]
    # Correction on resolved_tss:’tolist()’ on numpy.datetime64 is fine, but here we are in a list comprehension.
    # Just use the numpy datetime64 and it will be converted to python datetime by duckdb
    resolved_tss = [
        (np.datetime64(bts) + np.timedelta64(int(off), 's')) if prob > 0.2 else None 
        for off, prob in zip(resolved_offsets, resolved_probs)
    ]
    
    selected_priorities = np.take(priorities, priority_indices).tolist()
    selected_cats = np.take(ticket_cats, cat_indices).tolist()
    
    csat_values = [
        int(score) if prob > 0.3 else None 
        for score, prob in zip(csat_scores, csat_probs)
    ]
    is_resolved = (is_resolved_probs > 0.2).tolist()
    
    return list(zip(ticket_ids, selected_account_ids, created_tss, resolved_tss, selected_priorities, selected_cats, csat_values, is_resolved))


def main():
    sf = float(sys.argv[1]) if len(sys.argv) > 1 else 1.0
    sf_adj = sf * 800.0
    NAC = max(10, int(500 * sf_adj))
    NSB = max(10, int(700 * sf_adj))
    NEV = max(100, int(50000 * sf_adj))
    NFU = max(20, int(5000 * sf_adj))
    NST = max(10, int(2000 * sf_adj))


    os.makedirs("data", exist_ok=True)
    con = duckdb.connect("data/warehouse.duckdb")

    con.execute("""
    DROP TABLE IF EXISTS support_tickets; DROP TABLE IF EXISTS feature_usage;
    DROP TABLE IF EXISTS events; DROP TABLE IF EXISTS subscriptions;
    DROP TABLE IF EXISTS accounts;
    CREATE TABLE accounts(account_id INTEGER PRIMARY KEY, name VARCHAR,
        industry VARCHAR, country VARCHAR, arr DECIMAL(12,2),
        created_date DATE, csm_id INTEGER, health_score TINYINT);
    CREATE TABLE subscriptions(sub_id INTEGER PRIMARY KEY, account_id INTEGER,
        plan VARCHAR, seats INTEGER, mrr DECIMAL(10,2),
        start_date DATE, end_date DATE, is_active BOOLEAN, renewal_date DATE);
    CREATE TABLE events(event_id BIGINT PRIMARY KEY, account_id INTEGER,
        user_id INTEGER, event_type VARCHAR, event_ts TIMESTAMP,
        session_id VARCHAR, platform VARCHAR);
    CREATE TABLE feature_usage(fu_id INTEGER PRIMARY KEY, account_id INTEGER,
        feature_name VARCHAR, usage_date DATE, usage_count INTEGER);
    CREATE TABLE support_tickets(ticket_id INTEGER PRIMARY KEY, account_id INTEGER,
        created_ts TIMESTAMP, resolved_ts TIMESTAMP, priority VARCHAR,
        category VARCHAR, csat_score TINYINT, is_resolved BOOLEAN);
    """)

    base = date(2022, 1, 1)
    bts = datetime(2022, 1, 1)
    industries = ["fintech", "healthtech", "edtech", "ecommerce", "manufacturing", "media"]
    plans = ["starter", "growth", "enterprise", "enterprise_plus"]
    etypes = ["login", "page_view", "feature_click", "export", "api_call", "report_view"]
    features = ["dashboard", "reports", "api", "integrations", "automations", "analytics", "exports"]
    priorities = ["low", "medium", "high", "critical"]
    ticket_cats = ["billing", "technical", "feature_request", "onboarding", "other"]

    cpu_count = get_worker_count()
    progress = GenerationProgress("p06_saas", 5)
    with ProcessPoolExecutor(max_workers=cpu_count) as executor:
        progress.advance("accounts")
        batched_insert(con, "accounts", ['account_id', 'name', 'industry', 'country', 'arr', 'created_date', 'csm_id', 'health_score'],
                       run_parallel(executor, generate_accounts_chunk, NAC, industries, base))
        progress.advance("subscriptions")
        batched_insert(con, "subscriptions", ['sub_id', 'account_id', 'plan', 'seats', 'mrr', 'start_date', 'end_date', 'is_active', 'renewal_date'],
                       run_parallel(executor, generate_subscriptions_chunk, NSB, NAC, plans, base))
        progress.advance("events")
        batched_insert(con, "events", ['event_id', 'account_id', 'user_id', 'event_type', 'event_ts', 'session_id', 'platform'],
                       run_parallel(executor, generate_events_chunk, NEV, NAC, etypes, bts))
        progress.advance("feature_usage")
        batched_insert(con, "feature_usage", ['fu_id', 'account_id', 'feature_name', 'usage_date', 'usage_count'],
                       run_parallel(executor, generate_feature_usage_chunk, NFU, NAC, features, base))
        progress.advance("support_tickets")
        batched_insert(con, "support_tickets", ['ticket_id', 'account_id', 'created_ts', 'resolved_ts', 'priority', 'category', 'csat_score', 'is_resolved'],
                       run_parallel(executor, generate_support_tickets_chunk, NST, NAC, bts, priorities, ticket_cats))


    con.close()
    print_generation_summary(
        "p06_saas",
        sf,
        {
            "accounts": NAC,
            "subscriptions": NSB,
            "events": NEV,
            "feature_usage": NFU,
            "support_tickets": NST,
        },
    )

if __name__ == "__main__":
    main()

File: dbgen-rust/src/p06_saas.rs
use std::sync::Arc;

use chrono::{Duration, NaiveDate, NaiveDateTime};
use duckdb::arrow::array::{
    BooleanArray, Date32Array, Date64Array, Float64Array, Int32Array, Int64Array, Int8Array,
    StringArray, TimestampMillisecondArray, TimestampSecondArray,
};
use duckdb::arrow::datatypes::{
    ArrowTimestampType, Date32Type, Date64Type, TimestampMillisecondType, TimestampSecondType,
};
use duckdb::{Connection, DuckdbConnectionManager};
use indicatif::{ProgressBar, ProgressStyle};
use r2d2::Pool;
use rand::prelude::*;
use rand::rngs::SmallRng;

pub fn run(sf: f64, pool: &mut Pool<DuckdbConnectionManager>) -> duckdb::Result<()> {
    let sf_adj = sf * 800.0;
    let nac = (500.0 * sf_adj).max(10.0) as usize;
    let nsb = (700.0 * sf_adj).max(10.0) as usize;
    let nev = (50000.0 * sf_adj).max(100.0) as usize;
    let nfu = (5000.0 * sf_adj).max(20.0) as usize;
    let nst = (2000.0 * sf_adj).max(10.0) as usize;

    let con = &pool.get().expect("couldn't get connection");

    con.execute_batch(
        "DROP TABLE IF EXISTS support_tickets; DROP TABLE IF EXISTS feature_usage;
         DROP TABLE IF EXISTS events; DROP TABLE IF EXISTS subscriptions;
         DROP TABLE IF EXISTS accounts;
         CREATE TABLE accounts(account_id INTEGER PRIMARY KEY, name VARCHAR,
             industry VARCHAR, country VARCHAR, arr DECIMAL(12,2),
             created_date DATE, csm_id INTEGER, health_score TINYINT);
         CREATE TABLE subscriptions(sub_id INTEGER PRIMARY KEY, account_id INTEGER,
             plan VARCHAR, seats INTEGER, mrr DECIMAL(10,2),
             start_date DATE, end_date DATE, is_active BOOLEAN, renewal_date DATE);
         CREATE TABLE events(event_id BIGINT PRIMARY KEY, account_id INTEGER,
             user_id INTEGER, event_type VARCHAR, event_ts TIMESTAMP,
             session_id VARCHAR, platform VARCHAR);
         CREATE TABLE feature_usage(fu_id INTEGER PRIMARY KEY, account_id INTEGER,
             feature_name VARCHAR, usage_date DATE, usage_count INTEGER);
         CREATE TABLE support_tickets(ticket_id INTEGER PRIMARY KEY, account_id INTEGER,
             created_ts TIMESTAMP, resolved_ts TIMESTAMP, priority VARCHAR,
             category VARCHAR, csat_score TINYINT, is_resolved BOOLEAN);",
    )?;

    let base_date = NaiveDate::from_ymd_opt(2022, 1, 1).unwrap();
    let base_ts = base_date.and_hms_opt(0, 0, 0).unwrap();
    let industries = [
        "fintech",
        "healthtech",
        "edtech",
        "ecommerce",
        "manufacturing",
        "media",
    ];
    let plans = ["starter", "growth", "enterprise", "enterprise_plus"];
    let etypes = [
        "login",
        "page_view",
        "feature_click",
        "export",
        "api_call",
        "report_view",
    ];
    let features = [
        "dashboard",
        "reports",
        "api",
        "integrations",
        "automations",
        "analytics",
        "exports",
    ];
    let priorities = ["low", "medium", "high", "critical"];
    let ticket_cats = [
        "billing",
        "technical",
        "feature_request",
        "onboarding",
        "other",
    ];
    let countries = ["US", "UK", "DE", "Fregister_arrowR", "CA", "AU"];
    let platforms = ["web", "mobile", "api"];

    let pb = ProgressBar::new(5);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} {msg}")
            .unwrap(),
    );

    // 1. Accounts
    crate::generate_table(
        pool,
        "accounts",
        nac,
        &pb,
        "Generating accounts...",
        |start, end| {
            let mut rng = SmallRng::seed_from_u64(start as u64);
            let i: Vec<i64> = (start..end).map(|i| i as i64).collect();
            let name: Vec<String> = (start..end).map(|i| format!("Account {}", i)).collect();
            let industry: Vec<&str> = (start..end)
                .map(|_| industries[rng.gen_range(0..industries.len())])
                .collect();
            let country: Vec<&str> = (start..end)
                .map(|_| countries[rng.gen_range(0..countries.len())])
                .collect();
            let arr: Vec<f64> = (start..end)
                .map(|_| ((rng.gen_range(5000.0..500000.0) * 100.0) as f64).round() / 100.0)
                .collect();
            let created: Vec<i32> = (start..end)
                .map(|_| {
                    Date32Type::from_naive_date(base_date + Duration::days(rng.gen_range(0..701)))
                })
                .collect();
            let csm: Vec<i32> = (start..end).map(|_| rng.gen_range(1..21)).collect();
            let health: Vec<i8> = (start..end).map(|_| rng.gen_range(1..=100) as i8).collect();
            vec![
                Arc::new(Int64Array::from(i)),
                Arc::new(StringArray::from(name)),
                Arc::new(StringArray::from(industry)),
                Arc::new(StringArray::from(country)),
                Arc::new(Float64Array::from(arr)),
                Arc::new(Date32Array::from(created)),
                Arc::new(Int32Array::from(csm)),
                Arc::new(Int8Array::from(health)),
            ]
        },
    )?;

    // 2. Subscriptions
    crate::generate_table(
        pool,
        "subscriptions",
        nsb,
        &pb,
        "Generating subscriptions...",
        |start, end| {
            let mut rng = SmallRng::seed_from_u64(start as u64);
            let i: Vec<i64> = (start..end).map(|i| i as i64).collect();
            let acc_id: Vec<i32> = (start..end)
                .map(|_| rng.gen_range(1..=nac) as i32)
                .collect();
            let plan: Vec<&str> = (start..end)
                .map(|_| plans[rng.gen_range(0..plans.len())])
                .collect();
            let seats: Vec<i32> = (start..end).map(|_| rng.gen_range(1..201)).collect();
            let mrr: Vec<f64> = (start..end)
                .map(|_| ((rng.gen_range(99.0..9999.0) * 100.0) as f64).round() / 100.0)
                .collect();
            let start_date: Vec<i32> = (start..end)
                .map(|_| {
                    Date32Type::from_naive_date(base_date + Duration::days(rng.gen_range(0..601)))
                })
                .collect();
            let end_date: Vec<i32> = start_date
                .iter()
                .map(|s| {
                    Date32Type::from_naive_date(
                        Date32Type::to_naive_date_opt(*s).unwrap() + Duration::days(365),
                    )
                })
                .collect();
            let active: Vec<bool> = (start..end).map(|_| rng.gen_bool(0.9)).collect();
            vec![
                Arc::new(Int64Array::from(i)),
                Arc::new(Int32Array::from(acc_id)),
                Arc::new(StringArray::from(plan)),
                Arc::new(Int32Array::from(seats)),
                Arc::new(Float64Array::from(mrr)),
                Arc::new(Date32Array::from(start_date)),
                Arc::new(Date32Array::from(end_date.clone())),
                Arc::new(BooleanArray::from(active)),
                Arc::new(Date32Array::from(end_date)),
            ]
        },
    )?;

    // 3. Events
    crate::generate_table(
        pool,
        "events",
        nev,
        &pb,
        "Generating events...",
        |start, end| {
            let mut rng = SmallRng::seed_from_u64(start as u64);
            let i: Vec<i64> = (start..end).map(|i| i as i64).collect();
            let acc_id: Vec<i32> = (start..end)
                .map(|_| rng.gen_range(1..=nac) as i32)
                .collect();
            let user_id: Vec<i32> = (start..end)
                .map(|_| rng.gen_range(1..=nac * 5) as i32)
                .collect();
            let etype: Vec<&str> = (start..end)
                .map(|_| etypes[rng.gen_range(0..etypes.len())])
                .collect();
            let ts: Vec<i64> = (start..end)
                .map(|_| {
                    TimestampMillisecondType::from_naive_datetime(
                        base_ts + Duration::seconds(rng.gen_range(0..700 * 86400)),
                        None,
                    )
                    .unwrap()
                })
                .collect();
            let session: Vec<String> = (start..end)
                .map(|_| format!("sess_{}", rng.gen_range(1..nac * 20)))
                .collect();
            let platform: Vec<&str> = (start..end)
                .map(|_| platforms[rng.gen_range(0..platforms.len())])
                .collect();
            vec![
                Arc::new(Int64Array::from(i)),
                Arc::new(Int32Array::from(acc_id)),
                Arc::new(Int32Array::from(user_id)),
                Arc::new(StringArray::from(etype)),
                Arc::new(TimestampMillisecondArray::from(ts)),
                Arc::new(StringArray::from(session)),
                Arc::new(StringArray::from(platform)),
            ]
        },
    )?;

    // 4. Feature Usage
    crate::generate_table(
        pool,
        "feature_usage",
        nfu,
        &pb,
        "Generating feature usage...",
        |start, end| {
            let mut rng = SmallRng::seed_from_u64(start as u64);
            let i: Vec<i64> = (start..end).map(|i| i as i64).collect();
            let acc_id: Vec<i32> = (start..end)
                .map(|_| rng.gen_range(1..=nac) as i32)
                .collect();
            let feature: Vec<&str> = (start..end)
                .map(|_| features[rng.gen_range(0..features.len())])
                .collect();
            let date: Vec<i32> = (start..end)
                .map(|_| {
                    Date32Type::from_naive_date(base_date + Duration::days(rng.gen_range(0..701)))
                })
                .collect();
            let count: Vec<i32> = (start..end).map(|_| rng.gen_range(1..1001)).collect();
            vec![
                Arc::new(Int64Array::from(i)),
                Arc::new(Int32Array::from(acc_id)),
                Arc::new(StringArray::from(feature)),
                Arc::new(Date32Array::from(date)),
                Arc::new(Int32Array::from(count)),
            ]
        },
    )?;

    // 5. Support Tickets
    crate::generate_table(
        pool,
        "support_tickets",
        nst,
        &pb,
        "Generating support tickets...",
        |start, end| {
            let mut rng = SmallRng::seed_from_u64(start as u64);
            let i: Vec<i64> = (start..end).map(|i| i as i64).collect();
            let acc_id: Vec<i32> = (start..end)
                .map(|_| rng.gen_range(1..=nac) as i32)
                .collect();
            let created: Vec<i64> = (start..end)
                .map(|_| {
                    TimestampMillisecondType::from_naive_datetime(
                        base_ts + Duration::seconds(rng.gen_range(0..700 * 86400)),
                        None,
                    )
                    .unwrap()
                })
                .collect();
            let resolved: Vec<Option<i64>> = (start..end)
                .map(|_| {
                    if rng.gen_bool(0.8) {
                        Some(
                            TimestampMillisecondType::from_naive_datetime(
                                base_ts + Duration::seconds(rng.gen_range(0..700 * 86400)),
                                None,
                            )
                            .unwrap(),
                        )
                    } else {
                        None
                    }
                })
                .collect();
            let priority: Vec<&str> = (start..end)
                .map(|_| priorities[rng.gen_range(0..priorities.len())])
                .collect();
            let cat: Vec<&str> = (start..end)
                .map(|_| ticket_cats[rng.gen_range(0..ticket_cats.len())])
                .collect();
            let csat: Vec<Option<i8>> = (start..end)
                .map(|_| {
                    if rng.gen_bool(0.7) {
                        Some(rng.gen_range(1..6) as i8)
                    } else {
                        None
                    }
                })
                .collect();
            let is_resolved: Vec<bool> = (start..end).map(|_| rng.gen_bool(0.8)).collect();
            vec![
                Arc::new(Int64Array::from(i)),
                Arc::new(Int32Array::from(acc_id)),
                Arc::new(TimestampMillisecondArray::from(created)),
                Arc::new(TimestampMillisecondArray::from(resolved)),
                Arc::new(StringArray::from(priority)),
                Arc::new(StringArray::from(cat)),
                Arc::new(Int8Array::from(csat)),
                Arc::new(BooleanArray::from(is_resolved)),
            ]
        },
    )?;

    pb.finish_with_message("p06_saas complete");

    Ok(())
}
